collections/_article/rendering-acyclic-graph.md

```diff
@@ -0,0 +1,277 @@
+
+Whenever a resource is created, a new "tracker" structure is introduced to store the information relevant to the graph construction during command recording. The tracker holds references to which commands are writing or reading from the resource and modifies these lists accordingly as more commands are recorded. It also stores a "usage" variable that indicates what the current use of the resource is at the time of recording. Usages are both classified as "read" or "read-write" operations, and which one is used has strong implications for how dependencies between commands will be detected. For example, a command that reads from Resource A can be executed in parallel with another command that reads from Resource A, but that will not be valid if the other command can write to Resource A. In this case, a dependency is inserted between the two commands to ensure that the first command can finish reading the resource correctly before the next command modifies it.
+
+| ![](/storage/blog/acyclic-graph/resource-tracking.webp) |
+|:--:| 
+| *The tracker holds the current usage of a resource and determines whether it is necessary to perform a transition based on the type of command that references it.* |
+
+Textures also have a particular requirement: changing the usage implies a memory layout transition even if it's just for read-only operations. For example, there's different layouts for a texture being used as the source of a copy operation and for a texture being used for sampling in a shader. While this distinction might not necessarily be true at the hardware level, it is actually possible to witness texture corruption if these transitions are not performed correctly depending on the GPU's architecture (AMD is really good for testing these out!). Therefore, any change in usage when textures are involved is usually considered a write operation as most of them require a particular layout. This introduces some dependencies between commands that might not be very obvious but are completely required for the operations to work correctly: continuing with the previous example, it's not possible to use the optimal memory layout for copying a texture and sampling it in a shader in parallel, even if both are read-only operations.
+
+### Dependency tracking
+
+Since the graph construction is automatic and there's no input from the programmer in how the adjacency lists of the graph must be built, it's up to Rendering Device to use the resource trackers to figure out the dependencies between commands. While the final implementation is a bit more complex due to the introduction of texture "slices" (which is covered in another chapter), the main idea behind the algorithm is pretty straightforward.
```

Comment:
```suggestion
Since the graph construction is automatic and there's no input from the programmer in how the adjacency lists of the graph must be built, it's up to RenderingDevice to use the resource trackers to figure out the dependencies between commands. While the final implementation is a bit more complex due to the introduction of texture "slices" (which is covered in another chapter), the main idea behind the algorithm is pretty straightforward.
```